themes/default/content/blog/next-level-iac-pulumi-automation-api/index.md

```diff
@@ -0,0 +1,352 @@
+
+And it is this capability that allows you to build your own Service API on top of Pulumi. We can now build a REST or gRPC API endpoint that, depending on the request parameters, will create, update, or delete a Pulumi stack. This API can then be exposed to the rest of the organization, allowing developers to interact with their infrastructure in a programmatic way.
+
+## How to build your own Service API with Pulumi Automation API?
+
+Let's take a look at how you can build your own Service API with Pulumi Automation API.
+
+For this purpose, we will build two basic REST services writen in [TypeScript](https://www.typescriptlang.org/) and in [Python](https://www.python.org/).
+
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript
+import express from "express"
+import * as dotevnv from "dotenv"
+import cors from "cors"
+import helmet from "helmet"
+import {fullyQualifiedStackName, RemoteWorkspace} from "@pulumi/pulumi/automation";
+
+dotevnv.config()
+
+if (!process.env.PORT) {
+    console.log(`No port value specified...`)
+}
+
+const PORT = parseInt(process.env.PORT as string, 10)
+
+const app = express()
+
+app.use(express.json())
+app.use(express.urlencoded({extended: true}))
+app.use(cors())
+app.use(helmet())
+
+
+interface ServicePayload {
+    name: string;
+    org: string;
+    project: string;
+    stack: string;
+    region: string;
+}
+
+async function setUpPulumiProgram(payload: ServicePayload) {
+    const org = payload.org;
+    const project = payload.project;
+    const stackName = fullyQualifiedStackName(org, project, payload.stack);
+    const awsRegion = payload.region;
+
+    return await RemoteWorkspace.createOrSelectStack({
+        stackName,
+        url: "https://github.com/pulumi/examples.git",
+        branch: "refs/heads/master",
+        projectPath: project,
+    }, {
+        envVars: {
+            AWS_REGION: awsRegion,
+            AWS_ACCESS_KEY_ID: process.env.AWS_ACCESS_KEY_ID ?? "",
+            AWS_SECRET_ACCESS_KEY: {secret: process.env.AWS_SECRET_ACCESS_KEY ?? ""},
+            AWS_SESSION_TOKEN: {secret: process.env.AWS_SESSION_TOKEN ?? ""},
+        },
+    });
+}
+
+app.post('/', async (req, res) => {
+    const payload: ServicePayload = req.body;
+
+    if ((payload.name) || !payload.org || !payload.project || !payload.stack || !payload.region) {
+        res.status(400).send('Name, org, project, stack and region are required');
+        return;
+    }
+
+    const stack = await setUpPulumiProgram(payload);
+
+    const upRes = await stack.up({onOutput: console.log});
```

Comment:
This is where the real magic happens -- but it's _super_ easy to miss that what you're doing here is running Pulumi among this sea of code. We should at least drop some comments to draw attention to the most relevant portions of the code (the import statement, the calls to run `up`, etc.). It'd also be awesome to see what it actually looks like (in terms of output) when it runs, just so the user who's skimming can see what it is the content of the post is actually doing.

Reply:
done!